Handle both API versions

IanMatthewHuff opened the pull request:
Change the Notebook Language service code to handle either API >=1.56 and <1.56 so that python insiders will work on current stable vscode. Also will port to the release branch so that python extension can release without having to wait for vscode 1.56 to release to stable.

IanMatthewHuff commented on src/client/jupyter/languageserver/notebookConcatDocument.ts:
```diff
@@ -113,10 +118,14 @@ export class NotebookConcatDocument implements TextDocument, IDisposable {
         const dir = path.dirname(notebook.uri.fsPath);
         // Note: Has to be different than the prefix for old notebook editor (HiddenFileFormat) so
         // that the caller doesn't remove diagnostics for this document.
+        this._notebook = new SafeNotebookDocument(notebook);
```
@DonJayamanne This was more of my argument for using the wrapper in this case. When we are done with this, we just delete SafeNotebookDocument and set notebook back to NotebookDocument. Won't have to track down and worry about the 20 or so places that we access it. And old and new API can just use the same functions on it.

IanMatthewHuff replied:
Note, I have some cleanup and testing to do before submitting for review. Old comments and whatnot.

joyceerhl commented on src/client/jupyter/languageserver/safeNotebookDocument.ts:
```diff
@@ -0,0 +1,80 @@
+// In the Python extension we often need to support different changing
+// VS Code api versions. This class adds a layer of indirection so that we
+// can handle changes to the NotebookDocument class in the API
+export class SafeNotebookDocument implements ISafeNotebookDocument {
```
Suggestion only: we can create a GitHub issue for deprecating this wrapper and reference it in the comment here so we don't forget to remove it.

IanMatthewHuff replied:
Agreed, if we move forward with this PR I'll add the issue. Just don't want to add it if we don't go with the wrapper class.

DonJayamanne replied:
Rather than using a new class you can use the Proxy object.
This way u dont need to implement each property or method.
Just override getCells, cellAt and cellCount

DonJayamanne replied:
But thats an optionally improvement, hence a non blocking susggestion.
Else LGTM

IanMatthewHuff replied:
Proxy totally looks like what I was searching for earlier today. I'm going to go ahead with this as the goal here is to just remove this code in a few releases. I filed this issue #5503 to remove it once 1.56 has been in stable for a release or two.

IanMatthewHuff commented:
@kimadeline For context, was discussing this with @karthiknadig in the release branch PR. I had updated python side code to handle the newest VS Code notebook API, but realized that this would cause issues with the release (since VS Code is not on 1.56 yet) and also this would currently break the small number of customers using Native Notebooks + Python insiders + Stable vscode. Instead of forcing python to move to 1.56 I've updated the class here to handle either the old API or the new API. If we approve this approach will port to the release branch as well.

IanMatthewHuff commented:
Tested building python extension from main in both insiders and stable. Insiders followed the new API path and LS features worked in native notebooks. Stable followed the old cells path and LS feature worked.
